Added watsonmarlow323 device for i22 ppump.

barnettwilliam opened the pull request:
Added Watson Marlow 323 pump device for use on I22 - draft PR for concept review.

This device depends on an updated support module interface (support/watson-marlow) that was made more consistent while making this change - release TBC.

### Instructions to reviewer on how to test:
1.  TBC
2. Confirm that device is operable 

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://github.com/DiamondLightSource/dodal/wiki/Device-Standards)
- [ ] If changing the API for a pre-existing device, ensure that any beamlines using this device have updated their Bluesky plans accordingly
- [ ] Have the connection tests for the relevant beamline(s) been run via `dodal connect ${BEAMLINE}`

callumforrester commented on src/dodal/devices/watsonmarlow323.py:
```diff
@@ -0,0 +1,53 @@
+     STOPPED = "STOP"
+     STARTED = "START"
+
+class WatsonMarlow323(StandardReadable):
```
Nit: I would prefer a name that gives some suggestion as to what it does, maybe:

```suggestion
class WatsonMarlow323Pump(StandardReadable):
```

barnettwilliam replied:
Renamed as suggested also updated filenames in 417aaa12897e1aad61c73db30d4f3d6e00974e29.

callumforrester commented on tests/devices/unit_tests/test_watsonmarlow323.py:
```diff
@@ -0,0 +1,42 @@
+        },
+    )
+
+async def assert_reading(
```
Should: There is a utility function in `ophyd-async` that does this, you can just import that

Example: https://github.com/DiamondLightSource/dodal/blob/main/tests/devices/unit_tests/test_undulator.py

barnettwilliam replied:
Updated to use the ophyd-async definition as suggested in 417aaa12897e1aad61c73db30d4f3d6e00974e29, `assert_reading()` is defined in `test_slits.py` the file I initially looked at when creating the test for this device.

callumforrester commented on src/dodal/devices/watsonmarlow323.py:
```diff
@@ -0,0 +1,53 @@
+     """Watson Marlow 323 Pump device"""
+     def __init__(self, prefix: str, name: str= "") -> None:
+          
+          self.enabled = epics_signal_r(WatsonMarlow323Enable, prefix + "DISABLE")
```
I haven't read the manual, but is this PV really read-only?

barnettwilliam replied:
Checking this its not read only so I've made it rw in 417aaa12897e1aad61c73db30d4f3d6e00974e29.

callumforrester commented on src/dodal/devices/watsonmarlow323.py:
```diff
@@ -0,0 +1,53 @@
+     DISABLED = "Disabled"
+     ENABLED = "Enabled"
+
+class WatsonMarlow323Direction(str, Enum):
+     CW = "CW"
+     CCW = "CCW"
```
Nit: More meaningful names would be good here, we need to retain the cryptic acronyms on the epics side, but the ophyd side can have long, descriptive names e.g.

```suggestion
class WatsonMarlow323Direction(str, Enum):
     CLOCKWISE = "CW"
     COUNTER_CLOCKWISE = "CCW"
```

barnettwilliam replied:
Updated names in 417aaa12897e1aad61c73db30d4f3d6e00974e29

callumforrester reviewed (changes requested):
Looking good, thanks @barnettwilliam 

I've made a few comments, mostly minor stuff.

You also need to rebase against `main` and fix the linting. 
You can run `ruff format` and `ruff check --fix` to do that.
Additionally, if you use vscode, I recommend copy/pasting in [this settings directory](https://github.com/DiamondLightSource/python-copier-template/tree/main/.vscode) (it's gitignored) to make it auto format etc.

barnettwilliam commented:
Thanks @callumforrester I've incorporated your suggested changes, rebased, and linted. Also, made `enabled` a readable signal based on @DiamondJoseph's comments following chat on Wed. 

 I just need to test this with an actual device to confirm it communicates correctly after which I'll undraft and submit.

DiamondJoseph commented on src/dodal/beamlines/i22.py:
```diff
@@ -309,3 +310,16 @@ def linkam(
     )
+
+
+@skip_device
```
Why is this device skipped for i22?

barnettwilliam replied:
I didn't want interfere with anything, re-enabled in 6d5ae3fb866bdad8de27966fc11605b6c41856cd

DiamondJoseph commented on src/dodal/beamlines/i22.py:
```diff
@@ -309,3 +310,16 @@ def linkam(
+
+
+@skip_device
+def ppump(
+    wait_for_connection: bool = True, fake_with_ophyd_sim: bool = False
+) -> WatsonMarlow323Pump:
+    return device_instantiation(
+        WatsonMarlow323Pump,
+        "ppump",
+        "-EA-PUMP-01:",
+        wait_for_connection,
+        fake_with_ophyd_sim,
+    )
```
Please also add to p38.py with `fake_with_ophyd_sim: bool = True`

DiamondJoseph commented on src/dodal/devices/watsonmarlow323_pump.py:
```diff
@@ -0,0 +1,61 @@
+    def __init__(self, prefix: str, name: str = "") -> None:
+        self.enabled = epics_signal_rw(
+            WatsonMarlow323PumpEnable,
+            read_pv=prefix + "DISABLE",
+            write_pv=prefix + "DISABLE",
```
write_pv isn't required if they're both the same.

barnettwilliam replied:
Fixed in 6d5ae3fb866bdad8de27966fc11605b6c41856cd

DiamondJoseph commented on src/dodal/devices/watsonmarlow323_pump.py:
```diff
@@ -0,0 +1,61 @@
+        self.speed = epics_signal_rw(
+            float, read_pv=prefix + "INFO:SPD", write_pv=prefix + "SET:SPD"
+        )
+        self.limit_high = epics_signal_rw(
+            float, read_pv=prefix + "SET:SPD.DRVH", write_pv=prefix + "SET:SPD.DRVH"
+        )
+        self.limit_low = epics_signal_rw(
+            float, read_pv=prefix + "SET:SPD.DRVL", write_pv=prefix + "SET:SPD.DRVL"
+        )
```
Limits (and units, precision...) will propagate with the speed signal once [this change is merged](https://github.com/bluesky/ophyd-async/pull/137), these signals shouldn't be required: as is, it's not clear what they are the limits of, but when the change is merged they'll become metadata on the speed signal.

barnettwilliam replied:
Ok I'll remove these when bluesky/ophyd-async#137 is merged

barnettwilliam replied:
Rebased and fixed in 84824ca3d5a6f0e30f8cc00fa40fec3cea06a2aa.

DiamondJoseph commented on src/dodal/devices/watsonmarlow323_pump.py:
```diff
@@ -0,0 +1,61 @@
+            float, read_pv=prefix + "SET:SPD.DRVL", write_pv=prefix + "SET:SPD.DRVL"
+        )
+
+        self.set_readable_signals(
+            read=[self.state, self.speed, self.direction],
+            config=[
+                self.enabled,
+                self.limit_high,
+                self.limit_low,
+            ],
+        )
```
Nit: This and the above signal creation can be replaced with the new ContextManager on StandardReadable:

```suggestion
        with self.add_children_as_readables():
          self.direction = epics_signal_rw(
            WatsonMarlow323PumpDirection,
            read_pv=prefix + "INFO:DIR",
            write_pv=prefix + "SET:DIR",
          )
          self.state = epics_signal_rw(
            WatsonMarlow323PumpState,
            read_pv=prefix + "INFO:RUN",
            write_pv=prefix + "SET:RUN",
          )
          self.speed = epics_signal_rw(
            float, read_pv=prefix + "INFO:SPD", write_pv=prefix + "SET:SPD"
          )
        with self.add_children_as_readables(ConfigSignal):
          self.enabled = epics_signal_rw(
            WatsonMarlow323PumpEnable,
            read_pv=prefix + "DISABLE")
```

barnettwilliam replied:
I did initially consider the using the the context manager as shown in the tutorial but it wasn't clear to me from the name of HintedSignal what it was and I couldn't find any documentation clarifying this. `read` and `config` made more sense and there were more existing examples using this format.

The next device I create I'll use the context manager.

callumforrester replied:
https://github.com/bluesky/ophyd-async/issues/349

DominicOram commented on src/dodal/beamlines/i22.py:
```diff
@@ -309,3 +310,16 @@ def linkam(
+
+
+@skip_device
+def ppump(
```
Is this deliberately double p? If so what does the first p stand for?

Tom-Willemsen replied:
Interjecting here, the double p matches what's on i22's synoptic (it feels like matching the synoptic name <-> bluesky name makes sense?).

The p is for peristaltic pump I think, as opposed to spump which is a syringe pump I believe.

DominicOram replied:
Can we put this in the docstring? Otherwise people will (rightly) think it looks like a typo

callumforrester replied:
Or could call the device `peristaltic_pump`

barnettwilliam replied:
Added docstring description in 6d5ae3fb866bdad8de27966fc11605b6c41856cd, I'll find out what the scientists would prefer as the device name.

barnettwilliam commented on src/dodal/beamlines/i22.py:
```diff
@@ -272,3 +273,15 @@ def oav(
+        "-EA-PUMP-01:",
+        wait_for_connection,
+        fake_with_ophyd_sim
+    )
```
Fixed in 6d5ae3fb866bdad8de27966fc11605b6c41856cd

barnettwilliam replied:
Oops, thanks corrected in d21684dc4365cdacb2c697721190951967505ecb

DiamondJoseph commented on src/dodal/devices/watsonmarlow323_pump.py:
```diff
@@ -0,0 +1,52 @@
+    """Watson Marlow 323 Peristaltic Pump device"""
+
+    def __init__(self, prefix: str, name: str = "") -> None:
+        self.enabled = epics_signal_rw(
+            WatsonMarlow323PumpEnable,
+            prefix + "DISABLE",
+        )
+
+        self.direction = epics_signal_rw(
+            WatsonMarlow323PumpDirection,
+            read_pv=prefix + "INFO:DIR",
+            write_pv=prefix + "SET:DIR",
+        )
+        self.state = epics_signal_rw(
+            WatsonMarlow323PumpState,
+            read_pv=prefix + "INFO:RUN",
+            write_pv=prefix + "SET:RUN",
+        )
+        self.speed = epics_signal_rw(
+            float, read_pv=prefix + "INFO:SPD", write_pv=prefix + "SET:SPD"
+        )
+
+        self.set_readable_signals(
+            read=[self.state, self.speed, self.direction],
+            config=[
+                self.enabled,
+            ],
+        )
+
```
nit: this could be done with the `add_children_as_readables` context, which is now the recommended way to do it [in the ophyd-async docs](https://blueskyproject.io/ophyd-async/main/how-to/make-a-simple-device.html#readable)

DiamondJoseph reviewed (approved):
nit, but otherwise looks good

dperl-dls commented:
@barnettwilliam Do you still plan to merge this? Since you have addressed Callum's comments and Joseph has approved this you are free to dismiss Callum's review (or you could ask him to take a look again)
